src/scheduler/worker.rs

```diff
@@ -54,9 +74,15 @@ impl<C: Context> Worker<C> {
         &self.scheduler
     }
 
+    /// # Safety
+    /// The user needs to guarantee that the type supplied here is the same type used to create this pointer.
     #[inline]
-    pub fn local(&mut self) -> &mut C::WorkerLocal {
-        self.local.as_mut().unwrap()
+    pub unsafe fn local<W: WorkerLocal>(&mut self) -> &mut W {
+        self.local.as_type::<W>()
```

Comment:
Probably we can remove this unsafe, and
* For debug build, use dynamic downcast (`std::any::Any`), to ensure the code is correct
* For release build, simply use this force downcast in a unsafe block.

Reply:
Do you mean we use `Any` instead of `WorkerLocalPtr` for the `Worker.local` for debug build? 

The reason I put `unsafe` for this is `local()` is called from plan specific code, and the caller needs to provide the same `CopyContext` as they used to initiate `GCWorker` - we cannot guarantee this in this method.

Reply:
I agree with Yi, although it would be nice to perform a type check for debug build to save people from debugging headaches.

Reply:
Currently `WorkerLocalPtr` does not carry type info so we cannot use it to check type. I could try to see if I can add another field only for debugging build, and use it to assert the type.

Reply:
For debug build, I added a field `ty: &'static str` for `WorkerLocalPtr` to save the ptr type from `std::any::type_name()`, and check against it in `as_type()`. This is only for debug build so performance doesnt matter.